### PR TITLE
add device scannning speed-ups done for sle11-sp3 (bnc #910479)

### PR DESCRIPTION
**Pull request by wfeldt**
The goal of this is to avoid external program calls as much as possible.

### CODE REVIEW

File: src/FileIO.pm [hunk 1]
```diff
@@ -112,7 +137,45 @@ sub WriteFile
...
+      $self->error("Failed to close $file: $!");
+      $ok = 0;
+    }
+  }
+  else {
+    $self->error("Failed to open $file: $!");
+    $ok = 0;
+  }
+
+  umask $saved_umask;
```
Comment (jreidinger) [inline src/FileIO.pm]: is it related to this fix? it looks like different fix for bug with different umask

File: src/Tools.pm [hunk 2]
```diff
@@ -401,31 +412,41 @@ DMRaid Devices look like:
 DMRaid Partitions look like:
 <strange name>_part\d
 
-
 =cut
 
-sub ReadDMRaidPartitions {
+# FIXME: this has to be done through Storage
```
Comment (jreidinger) [inline src/Tools.pm]: this FIXME is wrong, as Tools module is intended to be used during kernel update and its goal is to be completelly independent on yast and also it should not use storage.

Comment (jreidinger) [inline src/Tools.pm]: BTW actually requirement( and one of feature I implement) for JEOS is to depend only on `perl-base` and not whole `perl`

File: src/Tools.pm [hunk 3]
```diff
@@ -484,42 +524,26 @@ returns 1 if yes, 0 if no
...
-    }
+sub IsDMRaidSlave
+{
+  my $self = $lib_ref;
+  my $dev = shift;
 
-    my $majmin_disk = `stat -c "%t:%T" $disk`;
-    chomp($majmin_disk);
-    my @dmparts = ();
+  my $r = $self->{cache}{IsDMRaidSlave}{$dev};
```
Comment (jreidinger) [inline src/Tools.pm]: please use more descriptive variable name. It is quite hard to guess what `r` means here.

File: src/Tools.pm [hunk 4]
```diff
@@ -484,42 +524,26 @@ returns 1 if yes, 0 if no
...
-    my @dm_devs = qx{$dmsetup info -c --noheadings -o name | grep -v part};
-    chomp @dm_devs;
+    return $r;
+  }
 
-    if ($dm_devs[0] !~ /No devices found/) {
-        foreach my $dmdisk (@dm_devs) {
-            my @tables = qx{$dmsetup table '$dmdisk'};
-            chomp @tables;
+  my @d = glob Bootloader::Path::Prefix("/sys/block/$dev/holders/*");
```
Comment (jreidinger) [inline src/Tools.pm]: please use more descriptive variable name. It is quite hard to guess what `d` means here.

File: src/Tools.pm [hunk 5]
```diff
@@ -530,22 +554,24 @@ otherwise 0.
...
 =cut
 
-sub IsDMDevice {
-    my $dev = shift;
+sub IsDMDevice
+{
+  my $self = $lib_ref;
+  my $dev = shift;
+
+  my $r = $self->{cache}{IsDMDevice}{$dev};
```
Comment (jreidinger) [inline src/Tools.pm]: please use more descriptive value. I have trouble to recognize, that it is actually return value.

File: src/FileIO.pm (hunk 1 above)
Comment (wfeldt) [inline src/FileIO.pm]: No, the umask fix was in Core.pm.

This is basically a port of the changes from sle11 and aligns both code streams.

File: src/Tools.pm (hunk 2 above)
Comment (wfeldt) [inline src/Tools.pm]: the FIXME was from the sle11 code, removed

Comment (wfeldt) [inline src/Tools.pm]: The patch doesn't change the package dependencies. The perl dep comes from the
'encoding(utf8)'.

File: src/Tools.pm (hunk 3 above)
Comment (wfeldt) [inline src/Tools.pm]: As this is really a port and not new code, I'd rather leave it as is. But of course, you are right.

**Comment by jreidinger** [comment]
LGTM

**Pull request merged**